### PR TITLE
Optimize panel league stepper

### PR DESCRIPTION
**Pull request by frostburn**
Only perform computationally intensive operations when the state changes sufficiently.
Implement simple recursive matrix flood fill and use it to make chaining flag propagation less computationally expensive.

### CODE REVIEW

File: lib/panel-league/basic.js [hunk 1]
```diff
@@ -73,11 +73,17 @@ const handleSwapping = ((state) => {
...
     }
+    else {
+      state.dirty = true;
+    }
   });
 });
 
 // Match three or more similar blocks horizontally or vertically
 const findMatches = ((state, blocks) => {
+  if (!state.dirty) {
```
Comment (RauliL) [inline lib/panel-league/basic.js]: Local variable declarations should always come first.

File: lib/panel-league/basic.js [hunk 2]
```diff
@@ -338,15 +338,32 @@ const handleChaining = ((state) => {
   if (!state.dirty) {
     return;
   }
-  floodFill(state, (block, neighbour) => {
-    if (block.chaining || !block.matching) {
-      return false;
+  // Flood-fills are expensive so we convert into
+  // matrix form and back to make it faster.
+  const source = [];
+  const target = [];
```
Comment (RauliL) [inline lib/panel-league/basic.js]: Blank lines.

File: lib/panel-league/basic.js [hunk 3]
```diff
@@ -338,15 +338,32 @@ const handleChaining = ((state) => {
...
+  for (let i = 0; i < state.height; ++i) {
+    const s_row = [];
```
Comment (RauliL) [inline lib/panel-league/basic.js]: These variable names contain underscore and do not describe what they are.

File: lib/panel-league/basic.js [hunk 4]
```diff
@@ -338,15 +338,32 @@ const handleChaining = ((state) => {
...
+    const t_row = [];
```
Comment (RauliL) [inline lib/panel-league/basic.js]: Also, blank lines.

File: lib/panel-league/basic.js [hunk 5]
```diff
@@ -338,15 +338,32 @@ const handleChaining = ((state) => {
...
+    for (let j = 0; j < state.width; ++j) {
+      const index = j + i * state.width;
+      const block = state.blocks[index];
```
Comment (RauliL) [inline lib/panel-league/basic.js]: Blank lines.

File: lib/panel-league/basic.js [hunk 6]
```diff
@@ -338,15 +338,32 @@ const handleChaining = ((state) => {
...
-  });
+    source.push(s_row);
+    target.push(t_row);
+  }
+  matrixFloodFill(source, target);
+  for (let i = 0; i < state.height; ++i) {
+    for (let j = 0; j < state.width; ++j) {
+      if (source[i][j]) {
+        const index = j + i * state.width;
+        const block = state.blocks[index];
```
Comment (RauliL) [inline lib/panel-league/basic.js]: Blank lines.

File: lib/panel-league/util.js [hunk 7]
```diff
@@ -92,6 +92,32 @@ module.exports.floodFill = ((state, callback) => {
   }
 });
 
+function seedFill(source, target, x, y) {
+  const row = source[y];
```
Comment (RauliL) [inline lib/panel-league/util.js]: Blank lines.

File: lib/panel-league/util.js [hunk 8]
```diff
@@ -92,6 +92,32 @@ module.exports.floodFill = ((state, callback) => {
...
+  if (row && row[x] === false && target[y][x]) {
```
Comment (RauliL) [inline lib/panel-league/util.js]:
```JavaScript
!row[x]
```
should be enough.

File: lib/panel-league/util.js [hunk 9]
```diff
@@ -92,6 +92,32 @@ module.exports.floodFill = ((state, callback) => {
...
+    seedFill(source, target, x+1, y);
+    seedFill(source, target, x, y-1);
+    seedFill(source, target, x, y+1);
+  }
+}
+
+// Expand source boolean matrix into target matrix
+module.exports.matrixFloodFill = ((source, target) => {
+  for (let i = 0; i < source.length; ++i) {
+    const row = source[i];
```
Comment (RauliL) [inline lib/panel-league/util.js]: Blank lines.

File: lib/panel-league/util.js [hunk 10]
```diff
@@ -92,6 +92,32 @@ module.exports.floodFill = ((state, callback) => {
   }
 });
 
+function seedFill(source, target, x, y) {
+  const row = source[y];
+  if (row && row[x] === false && target[y][x]) {
+    source[y][x] = true;
+    seedFill(source, target, x-1, y);
```
Comment (RauliL) [inline lib/panel-league/util.js]: Poorly formatted operators. Operators should be separated from operands with whitespace.

File: lib/panel-league/util.js [hunk 11]
```diff
@@ -92,6 +92,32 @@ module.exports.floodFill = ((state, callback) => {
...
+    seedFill(source, target, x+1, y);
+    seedFill(source, target, x, y-1);
+    seedFill(source, target, x, y+1);
+  }
+}
+
+// Expand source boolean matrix into target matrix
+module.exports.matrixFloodFill = ((source, target) => {
```
Comment (RauliL) [inline lib/panel-league/util.js]: Why is this function wrapped in parenthesis?

File: lib/panel-league/util.js (hunk 8 above)
Comment (frostburn) [inline lib/panel-league/util.js]: It's not. This algorithm abuses the fact that `arr[-1] === undefined`.

File: lib/panel-league/util.js (hunk 11 above)
Comment (frostburn) [inline lib/panel-league/util.js]: The rest of them were too.

File: lib/panel-league/basic.js (hunk 1 above)
Comment (frostburn) [inline lib/panel-league/basic.js]: This function is really expensive. I'd really like to short-circuit as soon as possible with no assignment overhead.

File: lib/panel-league/util.js (hunk 11 above)
Comment (frostburn) [inline lib/panel-league/util.js]: Should be a separate commit or PR to unify export definitions.

**Comment by frostburn** [comment]
Blank lines added.
